didcomm/pack.py

```diff
@@ -0,0 +1,128 @@
+from didcomm.secrets.secrets_resolver import SecretsResolver
+
+
+@dataclass(frozen=True)
+class PackResult:
+    """
+    Result of pack operation.
+
+    Attributes:
+        packed_msg (str): a packed message as a JSON string ready to be forwarded to the returned 'service_endpoint'
+        service_endpoint (str): an optional service endpoint to be used to transport the 'packed_msg'.
+    """
+    packed_msg: JSON
+    service_endpoint: Optional[str]
+
+
+@dataclass(frozen=True)
+class PackConfig:
+    """
+    Pack configuration.
+
+    Default config performs repudiable authentication encryption (auth_crypt)
+    and prepares a message ready to be forwarded to the returned endpoint (via Forward protocol).
+
+    Attributes:
+        secrets_resolver (SecretsResolver): an optional secrets resolver that can override a default secrets resolver
+        registered by 'register_default_secrets_resolver'
+
+        did_resolver (DIDResolver): an optional DID Doc resolver that can override a default DID Doc resolver
+        registered by 'register_default_did_resolver'
+
+        enc_alg_auth (AuthCryptAlg): the encryption algorithm to be used for authentication encryption (auth_crypt).
+        `A256CBC_HS512_ECDH_1PU_A256KW` by default.
+
+        enc_alg_anon (AnonCryptAlg): the encryption algorithm to be used for anonymous encryption (anon_crypt).
+        `XC20P_ECDH_ES_A256KW` by default.
+
+        encryption (bool): whether the plaintext needs to be encrypted. True by default.
+        authentication (bool): whether the plaintext needs to be authenticated. True by default.
+        anonymous_sender (bool): whether the sender ID needs to be protected. False by default.
+
+        forward (bool):  whether the packed messages need to be wrapped into Forward messages to be sent to Mediators
+        as defined by the Forward protocol. True by default.
+    """
+    secrets_resolver: SecretsResolver = None
+    did_resolver: DIDResolver = None
+    enc_alg_auth: AuthCryptAlg = AuthCryptAlg.A256CBC_HS512_ECDH_1PU_A256KW
+    enc_alg_anon: AnonCryptAlg = AnonCryptAlg.XC20P_ECDH_ES_A256KW
+    encryption: bool = True
+    authentication: bool = True
+    anonymous_sender: bool = False
+    forward: bool = True
+
+
+@dataclass(frozen=True)
+class PackParameters:
+    """
+    Optional paraneters for pack.
+
+    Attributes:
+        forward_headers (PlaintextOptionalHeaders): if forward is enabled (true by default),
+        optional headers can be passed to the wrapping Forward messages.
+
+        sign_frm (DID_OR_DID_URL): if non-repudiation is needed, a DID or key ID to be used for signing must be specified.
+        Not required by default as repudiation is expected in most of the cases.
+    """
+    forward_headers: Optional[PlaintextOptionalHeaders] = None
+    sign_frm: Optional[DID_OR_DID_URL] = None
+
+
+async def pack(plaintext: Plaintext,
+               frm: DID_OR_DID_URL, to: DID_OR_DID_URL,
```

Comment:
quite nitpicky, but I've always hated reducing variable names in libraries, particularly if it's only for the sake of a single character! Not such a big deal for e.g. `auth`, but `frm` just seems needlessly opaque

Reply:
`from` is a keyword in Python, that's the reason why we can not use it as-is.

Reply:
Aha! My python noobiness showing through :)

Reply:
@ashcherbakov often used pattern for snake case languages is `from_`